### ISSUE TITLE
"File Not Found" shown when blocked redirect has "Content-Disposition: attachment;"

### ISSUE THREAD

**Issue by jsamuel** [RequestPolicy/requestpolicy#180]
```
imported trac ticket
created: 2011-02-05 13:24:49
reporter: eibwen
```

This has always been a fun one.  Firefox throws a 404 file not found error, request policy shows the yellow redirect bar at the top.  Allow the request, and the 404 file becomes (presumably) 302, and the file downloads as expected.

Happens a lot with CDN download mirrors...

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-02-05 13:51:05
author: justin
```

Sorry, I don't completely understand the situation you're describing. Is a site sending a 404 response status along with a Location header and the resulting behavior is not what you expect? If you remember a link where this can be repeated, I can take a look. Thanks.

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-02-11 22:56:08
author: eibwen
```

Sourceforge seems to always trigger this with their CDN setup.
Random example:

http://downloads.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz?r=http%3A%2F%2Fsourceforge.net%2Fprojects%2Fgnome-sudoku%2F&ts=1297493496&use_mirror=softlayer

When the page "loads" firefox shows a "Problem loading page" error (presumed 404), and request policy shows a yellow bar at the top indicating a blocked redirect.  After the user allows the redirect, the yellow bar disappears and the file downloads, but firefox stays on the "problem loading page" page.

There is a potential for users to miss out on their downloads if they notice the 404 and not the blocked redirect...

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-02-12 09:46:24
author: justin
```

Thanks for providing an example. I actually hadn't seen this behavior before.

Here are the more detailed steps to repeat this:
- Use strictness level of "full domain"
- Go to [http://downloads.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz?r=http%3A%2F%2Fsourceforge.net%2Fprojects%2Fgnome-sudoku%2F&ts=1297493496&use_mirror=softlayer this sourceforge page](the same as the above provided link).
- In the RP menu, "temporarily allow from sourceforge.net to a.fsdn.com".
- In the RP menu, "temporarily allow from sourceforget.net to downloads.sourceforge.net" (temporary so that it's easier to restart firefox to test this again).

After the last step above, there will be a blocked Location header. The most relevant snippet of the RP logs are:

{{{
[RequestPolicy] [WARNING] [HEADER_REDIRECT] Showing notification for blocked redirect. To http://softlayer.dl.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz from http://downloads.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz?r=&ts=1297530056&use_mirror=softlayer
[RequestPolicy] [DEBUG] [INTERNAL] Checking for blocked content from page about:neterror?e=fileNotFound&u=http%3A//downloads.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz%3Fr%3D%26ts%3D1297530056%26use_mirror%3Dsoftlayer&c=UTF-8&d=Firefox%20can%27t%20find%20the%20file%20at%20http%3A//downloads.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz%3Fr%3D%26ts%3D1297530056%26use_mirror%3Dsoftlayer.
}}}

The content of the request/response with the blocked redirect looks like this:

{{{
GET /project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz?r=&ts=1297530676&use_mirror=softlayer HTTP/1.1
Host: downloads.sourceforge.net
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:2.0b12pre) Gecko/20110212 Firefox/4.0b12pre
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,_/_;q=0.8
Accept-Language: en-us,en;q=0.5
Accept-Encoding: gzip, deflate
Accept-Charset: ISO-8859-1,utf-8;q=0.7,*;q=0.7
Keep-Alive: 115
Connection: keep-alive
Referer: http://sourceforge.net/projects/gnome-sudoku/files/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz/download?use_mirror=softlayer
DNT: 1

HTTP/1.1 302 Found
X-Powered-By: PHP/5.2.9
Content-Disposition: attachment; filename="gnome-sudoku-0.7.1.tar.gz"
Location: http://softlayer.dl.sourceforge.net/project/gnome-sudoku/gnome-sudoku/0.7.1/gnome-sudoku-0.7.1.tar.gz
Content-type: text/html
Content-Length: 0
Date: Sat, 12 Feb 2011 17:11:18 GMT
Server: lighttpd/1.4.26
}}}

I think the issue here is that when Firefox gets the response with the Location header stripped by RP, there's nothing else in the response to display to the user because there is no content (content length is zero).

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-02-12 10:58:07
author: justin
```

Fixed in r406. When there is a "Content-Disposition: attachment;" header in a response with a blocked header redirect, Firefox displays "about:neterror?e=fileNotFound&...". At the same time as that error is displayed, RP is displaying the blocked redirect notification. The solution in this patch is to strip the Content-Disposition header when it's an attachment and RP has already removed the Location or Refresh header.

The behavior with this patch applied is now the same for "Content-Disposition: attachment;" and "Content-Disposition: inline;". That is, a blank page is displayed by Firefox when RP is showing the blocked redirect notification.

It turned out that my initial hunch about "Content-Length: 0" was wrong. The content will be shown as long as Content-Disposition is not "attachment". In other words, if RP wanted to display a custom message in the content of the page, it could add this content in these cases. However, I'd like to avoid making changes to content whenever possible. The blank page is probably less confusing, anyways, because there's only one thing for the user to look at: the notification from RP.

I've updated the summary to reflect the diagnosed problem.
